### ISSUE TITLE
Document how to use mkl from `defaults` with conda-forge

### ISSUE THREAD

**Issue by ocefpaf** [conda-forge/conda-forge.github.io#938]
This came up as a confusing point during the Pytorch meeting today.

**Comment by scopatz** [comment]
We should also probably make an MKL package for conda-forge.  We have approval from the NF lawyers to do so

**Comment by jjhelmus** [comment]
From my limited testing it seems that `conda create -n test numpy "blas=*=mkl"` is required to select the mkl variants of the blas packages.  

`conda create -n test numpy mkl` creates an environment with mkl, libopenblas and the openblas variants of the blas packages.

**Comment by jakirkham** [comment]
That's right.

It is documented [here]( https://conda-forge.org/docs/maintainer/knowledge_base.html#blas ).

Perhaps this is a question of discoverability? @soumith, do you have suggestions?

**Comment by ocefpaf** [comment]
Maybe we only need a FAQ pointing to that.

**Comment by soumith** [comment]
what is the way for a package to specify that it needs other packages to install their MKL variants?

The confusion (which I probably didn't articulate well on the call) comes when NumPy is loaded with OpenBLAS and PyTorch is loaded with MKL, some users have reported that they get incorrect results with some blas functions such as `sdot` (I think something to do with the Fortran-ABI return type being inconsistent), in some environments.

So, if PyTorch declares a dependency on `mkl` as the BLAS, and someone installs PyTorch (let's say conda-forge's version of PyTorch), then can the numpy package be changed over to the `mkl` variant?

**Comment by isuruf** [comment]
> The confusion (which I probably didn't articulate well on the call) comes when NumPy is loaded with OpenBLAS and PyTorch is loaded with MKL, some users have reported that they get incorrect results with some blas functions such as sdot (I think something to do with the Fortran-ABI return type being inconsistent), in some environments.

You have to realize that `conda-forge` and `defaults` are two different channels and issues in one does not necessarily mean that they are present in the others. That's why when reporting issues we ask for the conda environment list to see whether it's a `conda-forge` issue or a `defaults` issue and I close issues when they don't.

As to the issue at hand, we are using the Gfortran-ABI in MKL, https://github.com/conda-forge/blas-feedstock/blob/master/recipe/libblas_mkl_activate.sh#L2 while defaults is using the Intel-ABI. This choice was done in conda-forge, because we are compiling against the reference implementation in Gfortran-ABI and we want to be able to switch between BLAS implementations without an issue. There were some issues that cropped up and were promptly fixed. For eg: https://github.com/scipy/scipy/pull/10591 . If you know any issue like this we can fix them, but they need to be communicated to us.

TL:DR; that issue doesn't exist in `conda-forge`.

> So, if PyTorch declares a dependency on mkl as the BLAS, and someone installs PyTorch (let's say conda-forge's version of PyTorch), then can the numpy package be changed over to the mkl variant?

Yes, you can certainly do that. Know that this is a hard choice atm. From a hard choice that means, the user can't switch back to openblas if pytorch is installed. (It's possible to make it a soft choice by using features. i.e. if pytorch is installed, the default blas is MKL, but the user can force it to be OpenBLAS. This will rely on a conda feature that anaconda, inc is deprecating though.)

**Comment by isuruf** [comment]
> what is the way for a package to specify that it needs other packages to install their MKL variants?

See https://conda-forge.org/docs/maintainer/knowledge_base.html#switching-blas-implementation
